---
title: "TLD Example Vignette ONT Duplex Data"
author: "Jake Reed"
date: '`r Sys.Date()`'
output:
  html_document: 
    theme: simplex
    highlight: tango
    toc: yes
  pdf_document:
    highlight: tango
    toc: yes
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
        echo = TRUE,
        collapse = TRUE,
        comment = "#>"
)
```

# 1. Installation

# 2. Showcase

## Import Functions
```{r source}
# Start timer
start_time <- Sys.time()

source("00-paths.R")
source(file.path(paths$bin, "fG-helper_functions.R"))
library(Biostrings)

# Create directories
for (dir in paths) dir.create(file.path(dir), 
                              showWarnings = FALSE)

# Sample name
sample_name <- "telogator_example"

# Create results directory
dir.create(file.path(paths$example_results, sample_name), 
           showWarnings = FALSE)

sample_file <- file.path("/home/jake/science_projects/telo/other_tools/telogator2/test_data/hg002-telreads_ont.fa.gz")
data_dir <- file.path(paths$example_data, "data", 
                      "telogator_example")

# Set width
options(width = 300)
```

## Process Fastq
This process is used in order to split files in order to support efficient parallelization. 
Ideally, you would set `target_ram_size_gb` * `threads` to be equal to something near the 
total RAM for your machine. The fastq file can be downloaded from the SRA database using
SRA toolkit. Here are the commands you would use to download the fastq file for this
experiment. 

<!-- ```{bash}
# Download SRA toolkit
wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/3.0.7/sratoolkit.3.0.7-ubuntu64.tar.gz
tar -xzf sratoolkit.3.0.7-ubuntu64.tar.gz
cd sratoolkit.3.0.7-ubuntu64/bin
data_dir=r file.path(paths$example_data, "data")
echo $data_dir

# Download fastq file
./prefetch --progress --resume yes --output-directory $data_dir SRR11008518

# Convert to fastq
./fastq-dump --split-3 -A $data_dir/SRR11008518/SRR11008518.sra -outdir $data_dir/split
``` -->

```{r Split_Fastq, eval = TRUE}
source(file.path(paths$bin, "f01-Fastq_Processing.R"))

# Split large fastq file into files that will occupy
# 2GB of RAM in order to take advantage of all processors
# (20) with given RAM (62GB).
if(!file.exists(list.files(file.path(dirname(sample_file), "split"), 
                           full.names = TRUE)[1])) {
   stats_fl <- auto_split(in_fastq = sample_file,
                          prefix = "HG002_ONT_dup_", 
                          threads = 18, 
                          extension = ".gz", 
                          target_ram_size_gb = 2)
                          
  # Save stats
  saveRDS(stats_fl, file.path(paths$example_results,
                                sample_name, 
                                "stats_fl.rds"))
}

# Test cmd for ONT
# split_fastq(in_fastq = sample_file, 
#             size = "50000", 
#             prefix = "HG002_ONT", 
#             threads = 18, 
#             out_dir = paste0(data_dir, "/split"), 
#             extension = ".gz")
```

## Process Reference Fastq
```{r process_reference, eval = TRUE}
## Switching denovo motif analysis to streme from meme suite, need 
## to add to dependencies
source(file.path(paths$bin, "f02-Reference_Processing.R"))
options(meme_bin = "~/meme/bin")
reference_file <- file.path(paths$example_data, "ref", "hg002v1.0.1.fasta.gz")
reference <- readDNAStringSet(reference_file, format = "fasta")

# Process reference file 
# # PB Settings
# motif_analysis <- telomere_motif(data_ref = reference,
#                                  reference_telo_length = 2000,
#                                  telo_motif_length = 6, 
#                                  number_of_motifs = 10,
#                                  number_of_repeats = 9, 
#                                  platform = "PB",
#                                  threads = 18)

# ONT Settings
motif_analysis <- telomere_motif(data_ref = reference,
                                 reference_telo_length = 2000,
                                 telo_motif_length = 6, 
                                 number_of_motifs = 10,
                                 number_of_repeats = 15, 
                                 platform = "ONT",
                                 threads = 18)

print(paste("Number of motifs found: ", length(motif_analysis$motifs)))
cat("\n")
print(paste("Grep pattern for pattern matching and initial search: ", 
            motif_analysis$grep_list))
## Here it appears with new method we can use 6 X motif for ONT, maybe 9 X motif for PB???
## Maybe it doesn't matter. 

# Save results
saveRDS(motif_analysis, file.path(paths$example_results,
        sample_name, "motif_analysis_results.rds"))

# Clean up
rm(list = c("reference_file", "reference"))
gc()
```

## Isolate telomere reads
```{r telo_search_cut, eval = TRUE}
motif_analysis <- readRDS(file.path(paths$example_results,
                                    sample_name,
                                    "motif_analysis_results.rds"))
                                    
# Source functions for pattern searching
#source(file.path(paths$bin, "f03-Pattern_searching_V2.R"))
source(file.path(paths$bin, "f03-Pattern_searching.R"))

# Run pattern searh on all split fastq files
#file_list <- list.files(file.path(data_dir, "split"), 
#                        full.names = TRUE)
file <- file.path("/home/jake/science_projects/telo/other_tools/telogator2/test_data/hg002-telreads_ont.fa.gz")

# # For testing
# file_list <- file_list[1:100]
# cut_dir <- file.path(data_dir, "cut")
# dir.create(cut_dir, showWarnings = FALSE)

# cut <- cut_reads_par(fastq_files = file_list, 
# #                     grep_list = motif_analysis$grep_list, 
#                      threads = 14, 
#                      out_dir = cut_dir,
#                      return = TRUE, 
#                      progress = TRUE)

# saveRDS(cut, file.path(paths$example_results,
#                       sample_name,
#                       "cut.rds"))

# names(cut)
# cc <- unlist(cut["cut_reads"])
# head(cc)

# Create output directory
out <- file.path(data_dir, "telo")
dir.create(out, showWarnings = FALSE)

# Search telomeres 
## Need to explain in documentation that threads should match number of files
search_res <- telo_search(fastq_files = file,
                              grep_list = motif_analysis$grep_list,
                              threads = 14,
                              out_dir = out, 
                              progress = TRUE)
knitr::kable(head(search_res$telomere_stats))

# Clean up
rm(list = c("file", "out"))
gc()

# Save telomere_stats
saveRDS(telomere_stats, file.path(paths$example_results,
                                  sample_name,
                                  "telomere_stats.rds"))
```


## Isolate telomere reads
```{r telo_search, eval = TRUE}
motif_analysis <- readRDS(file.path(paths$example_results,
                                    sample_name,
                                    "motif_analysis_results.rds"))
                                    
# Source functions for pattern searching
source(file.path(paths$bin, "f03-Pattern_searching.R"))

# Run pattern searh on all split fastq files
# file_list <- list.files(file.path(data_dir, "split"), 
#                         full.names = TRUE)

file_list <- file.path("/home/jake/science_projects/telo/other_tools/telogator2/test_data/hg002-telreads_ont.fa.gz")
# # For testing
# file_list <- file_list[1:100]

# Create output directory
out <- file.path(data_dir, "telo")
dir.create(out, showWarnings = FALSE)

# Search telomeres 
## Need to explain in documentation that threads should match number of files
telomere_stats <- telo_search(fastq_files = file_list,
                              grep_list = motif_analysis$grep_list,
                              threads = 14,
                              out_dir = out, 
                              progress = TRUE)
knitr::kable(head(telomere_stats))

# Clean up
#rm(list = c("file_list", "out"))
rm(list = c("file", "out"))
gc()

# Save telomere_stats
saveRDS(telomere_stats, file.path(paths$example_results,
                                  sample_name,
                                  "telomere_stats.rds"))
```

## Cut telomere reads into sliding windows
```{r telo_cut, eval = TRUE}
# telomere_stats <- readRDS(file.path(paths$example_results,
#                                     sample_name,
#                                     "telomere_stats.rds"))

# Get telomere file list
telomere_files <- list.files(file.path(data_dir, "telo"), 
                             full.names = TRUE)

# Source functions for cutting telomere reads into sliding windows
source(file.path(paths$bin, "f04-Sliding_window.R"))
dir.create(file.path(paths$example_results, sample_name,
                     "windows"), 
           showWarnings = FALSE)
window_out <- file.path(paths$example_results, sample_name, "windows")

# Temporary file list
out <- file.path(data_dir, "telo")

# Generate sliding windows
#window_files <- sliding(telomere_files = telomere_stats$file_list, 
window_files <- sliding(telomere_files = telomere_files, 
        window_length = 200, 
        environment = "linux",
        threads = 18,
        return_windows = FALSE,
        save_files = TRUE,
        out_dir = window_out,
        progress = TRUE, 
        compress = TRUE, 
        verbose = TRUE, 
        process_file_number = 4)

# Satus
cat("\nFinished processing all files and saving results to", window_out, "\n")
```

## Calculate frequency of telomere motifs in sliding windows
```{r telo_freq, eval = TRUE}
# Source functions for calculating telomere motif frequencies
source(file.path(paths$bin, "f05-Generate_telomere_frequencies.R"))

# Load in data 
motif_analysis <- readRDS(file.path(paths$example_results, 
                                    sample_name, 
                                    "motif_analysis_results.rds"))
motifs <- as.character(motif_analysis$motifs)

# Directories
win_dir <- file.path(paths$example_results, 
        sample_name,
        "windows")

# For loop over all telomere windows files
wins <- list.files(win_dir, full.names = TRUE)

dir.create(file.path(paths$example_results, 
                     sample_name, "frequencies"), 
           showWarnings = FALSE)

out_dir = file.path(paths$example_results, 
                    sample_name, 
                    "frequencies")
# Test
freqs <- frequencies(windows_files = wins, 
                     motifs = motifs, 
                     environment = "linux", 
                     parallel = TRUE, 
                     threads = 18, 
                     save_files = TRUE,
                     progress = TRUE,
                     out_dir = out_dir,
                     process_file_number = 0, 
                     verbose = FALSE)

# Save Frequencies
saveRDS(freqs, file.path(paths$example_results,
                         sample_name, 
                         "freqs.rds"))
cat("\nNumber of telomeres found: ", length(freqs), "\n")
cat("\nExample telomere frequencies:\n")
knitr::kable(head(freqs[[1]]))
cat("\n...")
knitr::kable(tail(freqs[[1]]))

# Clean up
rm(list = setdiff(ls(), c("freqs", 
                          "paths", 
                          "sample_name", 
                          "start_time")))
gc()

# Save frequencies
saveRDS(freqs, file.path(paths$example_results,
                                        sample_name,
                                        "motif_frequencies.rds"))
```

# Determine telomere length
## !-- Somewhere I need to add telomere name designation in this 
## analysis --!
```{r telomere_length, eval = TRUE}
# Load data
telomere_frequencies <- readRDS(file.path(paths$example_results,
                                          sample_name,
                                          "motif_frequencies.rds"))

# Source functions for determining telomere length
source(file.path(paths$bin, "f06-Telomere_length.R"))

# Threshold determination
thresh_df <- determine_threshold(telomere_list = telomere_frequencies,
                                 threads = 18, 
                                 sample_ratio = 0.025)

# Save threshold data
saveRDS(thresh_df, file.path(paths$example_results,
                             sample_name, 
                             "threshold_data.rds"))
```

# Explore data
## Inspect threshold data
Need to look into why telomere lengths aren't being calculated for some reads.
```{r inspect_threshold_data, eval = TRUE}
# Load data
telomere_frequencies <- readRDS(file.path(paths$example_results,
                                          sample_name,
                                          "motif_frequencies.rds"))
thresh_df <- readRDS(file.path(paths$example_results,
                               sample_name,
                               "threshold_data.rds"))

# Number of submitted telomeres
print(paste0("Number of submitted telomeres: ", length(telomere_frequencies), 
        "\n"))

# Debugging
na_indices <- which(is.na(thresh_df$start[[50]]$telomere_end))
misfits <- telomere_frequencies[na_indices]
# Test all indices are NA
print(paste0("Number of NA (misassigned) telomeres: ", length(misfits), 
        "\n"))

# Fraction misassigned at best threshold
print(paste0("Fraction misassigned at 50-50 threshold: ",
        length(misfits) / length(telomere_frequencies), "\n"))

# Look into first read problem
ref_lens <- thresh_df$start[[50]]
head(ref_lens, 50)

# Look at average length in telomeres
telomere_lengths <- ref_lens$telomere_length[!is.na(ref_lens$telomere_end)]
print(paste0("Mean telomere length at 50-50 without NA: ", mean(telomere_lengths), 
        "\n"))

# Create histogrma of telomere lengths
hist(telomere_lengths, breaks = sqrt(length(telomere_lengths)), 
     main = "Histogram of Telomere Lengths", 
     xlab = "Telomere Length (bp)", 
     ylab = "Frequency")

# Less than 2k length
ind_2k <- which(telomere_lengths < 2000 & 
                !is.na(telomere_lengths) &
                telomere_lengths > 0)
print(paste0("Number of telomeres less than 2kbp: ", length(ind_2k), "\n"))
less_2k <- telomere_lengths[ind_2k]
print(paste0("Mean of telomere lengths at 50-50 which were less than 2 kbp: ", 
        mean(less_2k), "\n"))
hist(less_2k, breaks = sqrt(length(less_2k)), 
     main = "Histogram of Telomere Lengths < 2kbp", 
     xlab = "Telomere Length (bp)", 
     ylab = "Frequency")

minus_2k <- telomere_lengths[which(telomere_lengths > 2000)]
print(paste0("Mean of telomere lengths at 50-50 which were greater than 2 kbp: ", 
        mean(minus_2k), "\n"))
hist(minus_2k, breaks = sqrt(length(minus_2k)), 
     main = "Histogram of Telomere Lengths > 2kbp", 
     xlab = "Telomere Length (bp)", 
     ylab = "Frequency")
```

## Delve into threshold data, format
This section is going to be focused on constructing a dataset which 
can be looked at more closely. Basically we need a dataset based on 
starting threshold which shows sensitivity while the ending threshold
shows telomere length.
```{r formatting, eval =  TRUE}
# Start sensitivity
start_number_of_telos <- unlist(lapply(thresh_df$start, function(x) {
  length(which(!is.na(x$telomere_end)))}))
sensitivity_df <- data.frame(start = seq(1, 100, 1), 
                             end = rep(5, 100), 
                             number_of_telos = start_number_of_telos)

# Telomere length
# End telomere length
telomere_lengths_end <- lapply(thresh_df$end, function(x) {
  vals <- x$telomere_length[!is.na(x$telomere_end)]
  x <- mean(vals)
  if(x != 0 & length(vals) > 1) {
    std_dev <- sd(vals)
  } else {
    std_dev <- 0
  }
  ret_frame <- data.frame(mean = x,
                          std_dev = std_dev)
  return(ret_frame)
})
end_df <- do.call(rbind, telomere_lengths_end)
telomere_length_df <- cbind(start = rep(20, 100), 
                            end = seq(1, 100, 1), 
                            end_df)

# Print data
print(paste0("\nStart sensitivity data:"))
knitr::kable(head(sensitivity_df))
print(paste0("\n..."))
knitr::kable(tail(sensitivity_df))
print(paste0("\n"))

print(paste0("\nTelomere length data:"))
knitr::kable(head(telomere_length_df))
print(paste0("\n..."))
knitr::kable(tail(telomere_length_df))
```

## Plot threshold data
```{r plot_threshold_data, eval = TRUE}
library(ggplot2)
library(cowplot)

# Plot sensitivity and telomere line graphs next to each other
sens_plot <- ggplot(sensitivity_df, aes(x = start, y = number_of_telos)) +
  geom_line() +
  geom_point() +
  labs(x = "Start Threshold", 
       y = "Number of Telomeres", 
       title = "Telomere Sensitivity") +
  theme_bw()

telomere_plot <- ggplot(telomere_length_df, aes(x = end, y = mean)) +
  geom_line() +
  geom_point() +
  labs(x = "End Threshold", 
       y = "Telomere Length", 
       title = "Telomere Length") +
  theme_bw()

plot_grid(sens_plot, telomere_plot, ncol = 2)

# Print best thresholds
best_start <- sensitivity_df$start[which.max(sensitivity_df$number_of_telos)]
print(paste0("\nBest start threshold: ", best_start, "\n"))
best_end <- telomere_length_df$end[which.max(telomere_length_df$mean)]
print(paste0("\nBest end threshold: ", best_end, "\n"))
```

## Determine optimal thresholds
For now run all telomeres with start threshold of
25 and end threshold of 25.
```{r check_lengths, eval = TRUE}
source(file.path(paths$bin, "f06-Telomere_length.R"))

# Load data
telomere_frequencies <- readRDS(file.path(paths$example_results,
                                          sample_name,
                                          "motif_frequencies.rds"))

# Run findtellength over all telomeres in parallel
library(parallel)
# Setup parallel environment
cl <- parallel::makeCluster(18)
parallel::clusterExport(cl, c("findTelLength"), 
                        envir = environment())
invisible(parallel::clusterEvalQ(cl, require(plyr)))

results <- invisible(parallel::parLapply(cl, telomere_frequencies, 
                                         findTelLength, st.thresh = 25, 
                                         en.thresh = 25))
parallel::stopCluster(cl)

results_df <- do.call(rbind.data.frame, results)

head(results_df)

# Save results
saveRDS(results_df, file.path(paths$example_results,
                              sample_name, 
                              "results_df.rds"))

## This is very interesting
# saveRDS(results_df, file.path(paths$example_results,
#                               sample_name, 
#                               "results_df_interesting.rds"))

# Look at average length in telomeres
telomere_lengths <- results_df$telomere_length[which(!is.na(results_df$telomere_end) & 
                                               results_df$telomere_length > 0)]

# Print fraction na
print(paste0("Fraction NA: ", sum(is.na(results_df$telomere_end)) / nrow(results_df)))

# Telomere update print(paste0)
print(paste0("Number of telomeres found: ", length(telomere_lengths), "\n"))
wo_na <- results_df[!is.na(results_df$telomere_end), ]
print(paste0("Mean telomere length at 20-15 without NA: ", mean(wo_na$telomere_length)))

# Create histogrma of telomere lengths
hist(telomere_lengths, breaks = sqrt(length(telomere_lengths)), 
     main = "Histogram of Telomere Lengths", 
     xlab = "Telomere Length (bp)", 
     ylab = "Frequency")
```

# Truncation and Mapping
## Truncate reads
```{r truncate, eval = TRUE}
source(file.path(paths$bin, "fG-helper_functions.R"))
source(file.path(paths$bin, "f07-Truncate_Map.R"))
# Load data
results_df <- readRDS(file.path(paths$example_results,
                                sample_name, 
                                "results_df.rds"))

# Remove NAs
results_df <- results_df[!is.na(results_df$telomere_end), ]
ls()
find_sizes()

data_dir <- file.path(paths$example_data, "data", 
                      "telogator_example")

telomeres_fl <- list.files(file.path(data_dir, "telo"), 
                           full.names = TRUE)
ref_file <- file.path(paths$example_data, "ref", "hg002v1.0.1.fasta.gz")

# Truncate telomeres
trunc_dir <- file.path(data_dir, "trunc")
dir.create(trunc_dir, showWarnings = FALSE)

# Truncate files
# This step needs to be multithreaded, it's very slow and not working correctly
fasta_fl <- truncate_files(list_of_telomere_files = telomeres_fl, 
               results_data_frame = results_df,
               out_dir = trunc_dir,
               write_files = TRUE,
               return = FALSE,
               progress = TRUE)

fasta_fl <- as.list(list.files(trunc_dir, 
    full.names = TRUE, 
    include.dirs = FALSE))

# Create combined fasta file
combined_trunc_dir <- file.path(trunc_dir, "combined_trunc")
dir.create(combined_trunc_dir, showWarnings = FALSE)

# Combine fasta files
combine_fasta(list_of_fasta = fasta_fl,
              write_files = TRUE,
              out_dir = combined_trunc_dir, 
              prefix = "combined_trunc")

# Map reads to reference
combined_fasta <- file.path(combined_trunc_dir, "combined_trunc.fa")
map_dir <- file.path(data_dir, "map")
dir.create(map_dir, showWarnings = FALSE)

# Map reads
mapped_output <- map(fasta = combined_fasta, 
                     results_data_frame = results_df,
                     reference_file = ref_file, 
                     preset_string = "map-ont",
                     out_dir = map_dir, 
                     return_mapped = TRUE,
                     threads = 18)

# Save mapped output
saveRDS(mapped_output, file.path(paths$example_results,
                                 sample_name, 
                                 "mapped_output.rds"))
```

# Plotting
```{r violin_plot, eval = TRUE, fig.width = 12, fig.height = 8}
# Load data
mapped_output <- readRDS(file.path(paths$example_results,
                                   sample_name,
                                   "mapped_output.rds"))

# Create violin plot of mapped_output with top and bottom panels
library(ggplot2)
library(cowplot)
library(extrafont)
loadfonts()

# Histogram with mean indicated on telomere length data in results_df and fill by chromEnd
hist <- ggplot(mapped_output[["results"]], aes(x = telomere_length)) +
  geom_histogram(binwidth = 100, aes(fill = chromEnd, y = after_stat(density)), alpha = 0.3) +
  geom_density(alpha = 0.6, aes(y = after_stat(density), fill = chromEnd)) +
  geom_vline(aes(xintercept = mean(telomere_length, na.rm = TRUE)), 
             color = "#ff0051", linetype = "dashed", linewidth = 1) +
  labs(x = "Telomere Length", 
       y = "Frequency", 
       title = "Histogram of Telomere Lengths") +
       theme(text = element_text(size = 22, family = "Arial"))


left <- mapped_output[["results"]][mapped_output[["results"]]$chromEnd == "5'", ]
ord <- order(as.numeric(stringr::str_extract(left$ref_name, "\\d+")))
left <- left[ord, ]
left$ref_name <- factor(left$ref_name, levels = unique(left$ref_name))
right <- mapped_output[["results"]][mapped_output[["results"]]$chromEnd == "3'", ]
ord <- order(as.numeric(stringr::str_extract(right$ref_name, "\\d+")))
right <- right[ord, ]
right$ref_name <- factor(right$ref_name, levels = unique(right$ref_name))
top <- ggplot(left, aes(x = ref_name, y = telomere_length / 1000, fill = chromEnd)) +
  geom_violin(drop = FALSE) +
  geom_boxplot(width = 0.1) +
  scale_fill_manual(values = c("#0091ff")) +
  labs(x = "Chromosome", 
    y = "Telomere Length (kb)", 
    title = "Telomere Lengths by Chromosome End 5'") +
    theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1)) +
    theme(text = element_text(size = 18, family = "Arial"))

bottom <- ggplot(right, aes(x = ref_name, y = telomere_length / 1000, fill = chromEnd)) +
  geom_violin(drop = FALSE) +
  geom_boxplot(width = 0.1) +
  scale_fill_manual(values = c("#ff00b7")) +
  labs(x = "Chromosome", 
    y = "Telomere Length (kb)", 
    title = "Telomere Lengths by Chromosome End 3'") +
  theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1)) +
  theme(text = element_text(size = 18, family = "Arial"))

grid <- plot_grid(top, bottom, ncol = 1)
save_plot(file.path(paths$example_results, sample_name, "violin_plot.png"), grid, base_width = 12, base_height = 6)
save_plot(file.path(paths$example_results, sample_name, "histogram.png"), hist, base_width = 12, base_height = 6)

end_time <- Sys.time()
tot <- end_time - start_time
print(paste0("Total time: ", tot))
target <- 19+54+1661+98+413+76+390+38+0
target <- target / 60
if(as.numeric(tot) < target) {
    print(paste0("Total time is less than target time: ", (target - tot), "YAHOO!!"))
    } else {
    print(paste0("Total time is greater than target time: ", (target - tot), "BOO!!"))
}
```

```{r empty}
library(kableExtra)
data <- mapped_output$results

kable(head(data, 15), "latex", booktabs = T) %>%
  kable_styling(latex_options = c("striped", "scale_down")) %>%
  as_image(width = 50, file = file.path(paths$example_results, 
                                        sample_name, "ouput_results.png"))
```

